dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/deploy/DefaultApplicationDeployer.java

```diff
@@ -782,14 +782,17 @@ public void prepareApplicationInstance(ModuleModel moduleModel) {
         // export MetricsService
         exportMetricsService();
 
-        if (moduleModel.getDeployer().hasRegistryInteraction()) {
+        boolean registryConfiged = CollectionUtils.isNotEmpty(
+                moduleModel.getApplicationModel().getApplicationConfigManager().getRegistries());
+
+        if (registryConfiged) {
```

Comment:
Comparing to `hasRegistryInteraction()`, the new condition missed to check the registry configuration that is restricted in ServiceConfig. So this change might fix the registration issue for users programmatically using API but probably will cause new issue for users exporting services the other way.

Reply:
> Comparing to `hasRegistryInteraction()`, the new condition missed to check the registry configuration that is restricted in ServiceConfig. So this change might fix the registration issue for users programmatically using API but probably will cause new issue for users exporting services the other way.

i think it should to start the registry schedule if there is a registry config in application config. because service export may happens after dubbo started, so the instance can't be registry.